File: src/main/java/main/audiohandlers/loaders/SearchResultLoader.java
package main.audiohandlers.loaders;

import com.sedmelluq.discord.lavaplayer.player.AudioLoadResultHandler;
import com.sedmelluq.discord.lavaplayer.tools.FriendlyException;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackInfo;
import main.audiohandlers.sources.spotify.SpotifySourceManager;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.component.interactions.selectionmenu.SelectionMenuBuilder;
import main.utils.json.themes.ThemesConfig;
import main.utils.locale.LocaleManager;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.channel.middleman.GuildMessageChannel;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.interactions.InteractionHook;
import net.dv8tion.jda.api.interactions.components.ActionRow;
import net.dv8tion.jda.api.interactions.components.buttons.ButtonStyle;
import net.dv8tion.jda.api.interactions.components.buttons.Button;
import net.dv8tion.jda.api.interactions.components.selections.SelectMenu;
import net.dv8tion.jda.internal.utils.tuple.Pair;

public class SearchResultLoader implements AudioLoadResultHandler {
    private final Guild guild;
    private final User searcher;
    private final String query;
    private final Message botMsg;
    private final InteractionHook interactionBotMsg;

    public SearchResultLoader(Guild guild, User searcher, String query, Message botMsg) {
        this.guild = guild;
        this.searcher = searcher;
        this.query = query;
        this.botMsg = botMsg;
        this.interactionBotMsg = null;
    }

    public SearchResultLoader(Guild guild, User searcher, String query, InteractionHook botMsg) {
        this.guild = guild;
        this.searcher = searcher;
        this.query = query;
        this.interactionBotMsg = botMsg;
        this.botMsg = null;
    }

    @Override
    public void trackLoaded(AudioTrack track) {
        throw new UnsupportedOperationException("This operation is not supported in the search result loader");
    }

    @Override
    public void playlistLoaded(AudioPlaylist playlist) {
        if (!playlist.isSearchResult())
            throw new UnsupportedOperationException("This operation is not supported in the search result loader");

        final var tracks = playlist.getTracks();
        final var localeManager = LocaleManager.getLocaleManager(guild);
        SelectionMenuBuilder selectionMenuBuilder = new SelectionMenuBuilder()
                .setName("searchresult:" + searcher.getId() + ":" + query.toLowerCase()
                        .replaceAll(" ", "%SPACE%"))
                .setPlaceHolder(localeManager.getMessage(RobertifyLocaleMessage.SearchMessages.SEARCH_MENU_PLACEHOLDER))
                .setRange(1, 1);

        final StringBuilder embedDescription = new StringBuilder();

        for (int i = 0; i < Math.min(10, tracks.size()); i++) {
            AudioTrackInfo info = tracks.get(i).getInfo();
            selectionMenuBuilder.addOption(info.title, info.identifier, null);
            embedDescription.append("**").append(i+1).append(".** - ").append(info.title).append(" by ")
                    .append(info.author).append(" [").append(GeneralUtils.formatTime(info.length))
                    .append("]").append("\n");
        }

        SelectMenu selectionMenu = selectionMenuBuilder.build();

        if (botMsg != null)
            botMsg.editMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, embedDescription.toString())
                            .setAuthor(localeManager.getMessage(RobertifyLocaleMessage.SearchMessages.SEARCH_EMBED_AUTHOR, Pair.of("{query}", query.replaceFirst(SpotifySourceManager.SEARCH_PREFIX, ""))), null, new ThemesConfig(guild).getTheme().getTransparent())
                            .setFooter(localeManager.getMessage(RobertifyLocaleMessage.SearchMessages.SEARCH_EMBED_FOOTER))
                            .build())
                    .setComponents(
                            ActionRow.of(selectionMenu),
                            ActionRow.of(Button.of(ButtonStyle.DANGER, "searchresult:end:" + searcher.getId(), localeManager.getMessage(RobertifyLocaleMessage.SearchMessages.SEARCH_END_INTERACTION)))
                    )
                    .queue();
        else interactionBotMsg.sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, embedDescription.toString())
                        .setAuthor(localeManager.getMessage(RobertifyLocaleMessage.SearchMessages.SEARCH_EMBED_AUTHOR, Pair.of("{query}", query.replaceFirst(SpotifySourceManager.SEARCH_PREFIX, ""))), null, new ThemesConfig(guild).getTheme().getTransparent())
                        .setFooter(localeManager.getMessage(RobertifyLocaleMessage.SearchMessages.SEARCH_EMBED_FOOTER))
                        .build())
                .setComponents(
                        ActionRow.of(selectionMenu),
                        ActionRow.of(Button.of(ButtonStyle.DANGER, "searchresult:end:" + searcher.getId(), localeManager.getMessage(RobertifyLocaleMessage.SearchMessages.SEARCH_END_INTERACTION)))
                )
                .setEphemeral(RobertifyEmbedUtils.getEphemeralState((GuildMessageChannel) interactionBotMsg.getInteraction().getMessageChannel()))
                .queue();
    }

    @Override
    public void noMatches() {
        botMsg.editMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.AudioLoaderMessages.NO_TRACK_FOUND, Pair.of("{query}", query.replaceFirst(SpotifySourceManager.SEARCH_PREFIX, ""))).build())
                .queue();
    }

    @Override
    public void loadFailed(FriendlyException exception) {

    }
}

File: src/main/kotlin/main/audiohandlers/loaders/SearchResultLoader.kt
package main.audiohandlers.loaders

import com.github.topi314.lavasrc.spotify.SpotifySourceManager
import dev.arbjerg.lavalink.client.player.Track
import dev.arbjerg.lavalink.client.player.TrackException
import dev.arbjerg.lavalink.protocol.v4.PlaylistInfo
import dev.minn.jda.ktx.interactions.components.danger
import dev.minn.jda.ktx.messages.send
import main.audiohandlers.GuildMusicManager
import main.utils.GeneralUtils
import main.utils.RobertifyEmbedUtils
import main.utils.component.interactions.selectionmenu.StringSelectMenuOption
import main.utils.component.interactions.selectionmenu.StringSelectionMenuBuilder
import main.utils.json.themes.ThemesConfig
import main.utils.locale.LocaleManager
import main.utils.locale.messages.AudioLoaderMessages
import main.utils.locale.messages.SearchMessages
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.entities.emoji.Emoji
import net.dv8tion.jda.api.interactions.InteractionHook
import net.dv8tion.jda.api.interactions.components.ActionRow

class SearchResultLoader(
    override val musicManager: GuildMusicManager,
    private val searcher: User,
    override val query: String,
    private val botMessage: InteractionHook
) : AudioLoader(musicManager.guild) {
    private val guild = musicManager.guild

    override fun onPlaylistLoad(playlist: List<Track>, playlistInfo: PlaylistInfo) {
        throw UnsupportedOperationException("This operation is not supported in the search result loader")
    }

    override fun onSearchResultLoad(results: List<Track>) {
        val localeManager = LocaleManager[guild]
        val embedDesc = StringBuilder()

        val options = results.subList(0, results.size.coerceAtMost(10)).mapIndexed { i, track ->
            val info = track.info
            val label = "${info.title} by ${info.author}"
            val safeLabel = label.substring(0, label.length.coerceAtMost(100))

            embedDesc.append("**${i + 1}.** - ${info.title} by ${info.author} [${GeneralUtils.formatTime(info.length)}]\n")

            StringSelectMenuOption(
                label = safeLabel,
                value = "https://open.spotify.com/track/${info.identifier}",
                emoji = Emoji.fromUnicode(GeneralUtils.parseNumEmoji(i + 1))
            )
        }

        val selectionMenu = StringSelectionMenuBuilder(
            _name = "searchresult:${searcher.id}:${query.lowercase().replace(" ", "%SPACE%")}",
            placeholder = localeManager.getMessage(SearchMessages.SEARCH_MENU_PLACEHOLDER),
            range = Pair(1, 1),
            _options = options
        ).build()

        botMessage.send(
            embeds = listOf(
                RobertifyEmbedUtils.embedMessage(guild, embedDesc.toString())
                    .setAuthor(
                        localeManager.getMessage(
                            SearchMessages.SEARCH_EMBED_AUTHOR,
                            Pair("{query}", query.replaceFirst(SpotifySourceManager.SEARCH_PREFIX, ""))
                        ),
                        null,
                        ThemesConfig(guild).getTheme().transparent
                    )
                    .setFooter(localeManager.getMessage(SearchMessages.SEARCH_EMBED_FOOTER))
                    .build()
            ),
            components = listOf(
                ActionRow.of(selectionMenu),
                ActionRow.of(
                    danger(
                        label = localeManager.getMessage(SearchMessages.SEARCH_END_INTERACTION),
                        id = "searchresult:end:${searcher.id}"
                    )
                )
            )
        ).queue()
    }

    override fun onTrackLoad(result: Track) {
        throw UnsupportedOperationException("This operation is not supported in the search result loader")
    }

    override fun onNoMatches() {
        botMessage.editOriginalEmbeds(
            RobertifyEmbedUtils.embedMessage(
                guild,
                AudioLoaderMessages.NO_TRACK_FOUND,
                Pair("{query}", query.replaceFirst(SpotifySourceManager.SEARCH_PREFIX, ""))
            ).build()
        )
            .queue()
    }

    override fun onException(exception: TrackException) {
        TODO("Not yet implemented")
    }
}